README.md
# Santi's Base64 Library

[![Build Status](https://github.com/santi100a/base64-lib/actions/workflows/main.yml/badge.svg)](https://github.com/santi100a/base64-lib/actions)
[![GitHub stars](https://img.shields.io/github/stars/santi100a/base64-lib.svg)](https://github.com/santi100a/base64-lib)
[![License](https://img.shields.io/github/license/santi100a/base64-lib.svg)](https://github.com/santi100a/base64-lib)
[![npm bundle size](https://img.shields.io/bundlephobia/min/@santi100/base64-lib)](https://bundlephobia.com/package/@santi100/base64-lib@latest)


- 🚀 Lightweight and fast[^](#disclaimers)
- 👴 ES3-compliant[*](#disclaimers)
- 💻 Portable between the browser and Node.js


## What's this?
This is a lightweight, fast library for encoding and decoding strings in base64 format.
It's portable between the browser and Node.js. In order to keep backwards compatibility with old browsers, this library can only encode and decode strings and buffers (if it's running in Node.js). This serves as a lightweight, portable replacement for the `btoa` and `atob` functions, and provides clearer names. 

## Contribute

Wanna contribute? [File an issue](https://github.com/santi100a/base64-lib/issues) or [pull request](https://github.com/santi100a/base64-lib/pulls)!
Make sure you follow the [contribution Code of Conduct](https://github.com/santi100a/base64-lib/blob/main/CODE_OF_CONDUCT.md).
## Installation
- Via NPM: `npm install @santi100/base64-lib`
- Via Yarn: `yarn add @santi100/base64-lib`
- Via PNPM: `pnpm install @santi100/base64-lib`

## API
- `function encode(input: string): string;` Encodes `input` to base64 format. It takes a string as an argument, and returns a base64 representation of `input`.
- `function decode(input: string): string;` Decodes `input` back to its original form. Must be a valid base64 string. It takes a valid base64 string, and returns the decoded value of `input`.

## My Goals
The above functions must:
1. Be able to handle strings that contain non-printable ASCII (the result of stringifying a buffer, for example).
2. Be fully ES3 compliant (no ArrayBuffer, Uint8Array or anything like that).
3. Use padding `=` signs.
4. Be fully portable between the browser and Node.js (no reliance in any browser- or Node-exclusive API).
5. Not modify the argument vector.

**If any of the above is not being satisfied, open a pull request or issue. I have no idea how to satisfy idea #1 reliably (see [Contribute](CONTRIBUTE.md) (for instructions) and [disclaimers](#disclaimers)😀).**

### Disclaimers
- **Keep in mind this library can only reliably handle plaintext for now.**
- **Hasn't been tested in an actual ES3 environment. Feel free to open an issue or pull request if you find any non-ES3 thing.*
- *^The source code is about 3 kilobytes.*

CHANGELOG.md
# Changelog

## Version 0.0.4
- Removed unnecessary JSDoc.
- Added support for Buffers without breaking ES3 backward compatibility (theoretically).

## Version 0.0.5
- Fixed security vulnerability (see [#1](https://github.com/santi100a/base64-lib/issues/1) for details).

tests/decode.test.js
describe('decode function tests', () => {
    const decode = require('../cjs').decode;
    test('it\'s a valid function', () => {
        expect(typeof decode)
            .toBe('function');
    });
    test('it can decode any string back to its original form', () => {
        expect(decode('aGVsbG8gYmVhcmVyIHRva2Vu'))
            .toBe('hello bearer token');
        expect(decode('VGhlIHF1aWNrIGJyb3duIGZveCBqdW1wcyBvdmVyIHRoZSBsYXp5IGRvZw=='))
            .toBe('The quick brown fox jumps over the lazy dog');
        expect(decode('VGhlIGVhcmx5IGJpcmQgY2F0Y2hlcyB0aGUgd29ybQ==')) 
            .toBe('The early bird catches the worm');
        expect(decode('VGhlIHBlbiBpcyBtaWdodGllciB0aGFuIHRoZSBzd29yZA=='))
            .toBe('The pen is mightier than the sword');
        expect(decode('QWxsJ3MgZmFpciBpbiBsb3ZlIGFuZCB3YXI='))
            .toBe('All\'s fair in love and war'); 
        expect(decode('V2hlcmUgdGhlcmUncyBzbW9rZSwgdGhlcmUncyBmaXJl'))
            .toBe('Where there\'s smoke, there\'s fire');
    });
    test('it can decode buffers', () => { 
        expect(decode(Buffer.from('aGVsbG8gYmVhcmVyIHRva2Vu')))
            .toBe('hello bearer token');
        expect(decode(Buffer.from('VGhlIHF1aWNrIGJyb3duIGZveCBqdW1wcyBvdmVyIHRoZSBsYXp5IGRvZw==')))
            .toBe('The quick brown fox jumps over the lazy dog');
        expect(decode(Buffer.from('VGhlIGVhcmx5IGJpcmQgY2F0Y2hlcyB0aGUgd29ybQ==')))
            .toBe('The early bird catches the worm');
        expect(decode(Buffer.from('VGhlIHBlbiBpcyBtaWdodGllciB0aGFuIHRoZSBzd29yZA==')))
            .toBe('The pen is mightier than the sword');
        expect(decode(Buffer.from('QWxsJ3MgZmFpciBpbiBsb3ZlIGFuZCB3YXI=')))
            .toBe('All\'s fair in love and war'); 
        expect(decode(Buffer.from('V2hlcmUgdGhlcmUncyBzbW9rZSwgdGhlcmUncyBmaXJl')))
            .toBe('Where there\'s smoke, there\'s fire');
    });
}); 

tests/encode.test.js
describe('encode function tests', () => {
    const encode = require('../cjs').encode;
    test('it\'s a valid function', () => {
        expect(typeof encode)
            .toBe('function');
    });
    test('it can encode any string into a valid format', () => {
        expect(encode('hello bearer token'))
            .toBe('aGVsbG8gYmVhcmVyIHRva2Vu');
        expect(encode('The quick brown fox jumps over the lazy dog'))
            .toBe('VGhlIHF1aWNrIGJyb3duIGZveCBqdW1wcyBvdmVyIHRoZSBsYXp5IGRvZw==');
        expect(encode('The early bird catches the worm'))
            .toBe('VGhlIGVhcmx5IGJpcmQgY2F0Y2hlcyB0aGUgd29ybQ==');
        expect(encode('The pen is mightier than the sword'))
            .toBe('VGhlIHBlbiBpcyBtaWdodGllciB0aGFuIHRoZSBzd29yZA==');
        expect(encode('All\'s fair in love and war'))
            .toBe('QWxsJ3MgZmFpciBpbiBsb3ZlIGFuZCB3YXI=');
        expect(encode('Where there\'s smoke, there\'s fire'))
            .toBe('V2hlcmUgdGhlcmUncyBzbW9rZSwgdGhlcmUncyBmaXJl');        
    });
    test('it can encode buffers', () => {
        expect(encode(Buffer.from('hello bearer token')))
            .toBe('aGVsbG8gYmVhcmVyIHRva2Vu');
        expect(encode(Buffer.from('The quick brown fox jumps over the lazy dog')))
            .toBe('VGhlIHF1aWNrIGJyb3duIGZveCBqdW1wcyBvdmVyIHRoZSBsYXp5IGRvZw==');
        expect(encode(Buffer.from('The early bird catches the worm')))
            .toBe('VGhlIGVhcmx5IGJpcmQgY2F0Y2hlcyB0aGUgd29ybQ==');
        expect(encode(Buffer.from('The pen is mightier than the sword')))
            .toBe('VGhlIHBlbiBpcyBtaWdodGllciB0aGFuIHRoZSBzd29yZA==');
        expect(encode(Buffer.from('All\'s fair in love and war')))
            .toBe('QWxsJ3MgZmFpciBpbiBsb3ZlIGFuZCB3YXI=');
        expect(encode(Buffer.from('Where there\'s smoke, there\'s fire')))
            .toBe('V2hlcmUgdGhlcmUncyBzbW9rZSwgdGhlcmUncyBmaXJl');        
    });
});

cjs/index.js
"use strict";
exports.__esModule = true;
exports.decode = exports.encode = void 0;
function createCoder() {
    return new /** @class */ (function () {
        function Coder(opts) {
            if (opts === void 0) { opts = {}; }
            this.opts = opts;
        }
        Coder.prototype.encode = function (input) {
            return encode(input, this.opts);
        };
        Coder.prototype.decode = function (input) {
            return decode(input, this.opts);
        };
        return Coder;
    }());
}
function map(array, fn) {
    var newValues = [];
    for (var _i = 0, array_1 = array; _i < array_1.length; _i++) {
        var item = array_1[_i];
        newValues.push(fn(item, array.indexOf(item), array));
    }
    return newValues;
}
var URL_SAFE_CHARS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_';
var STANDARD_CHARS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
var CHAR_LOOKUP = {};
for (var i = 0; i < URL_SAFE_CHARS.length; i++) {
    CHAR_LOOKUP[URL_SAFE_CHARS.charAt(i)] = i;
}
for (var i = 0; i < STANDARD_CHARS.length; i++) {
    CHAR_LOOKUP[STANDARD_CHARS.charAt(i)] = i;
}
/**
 *
 * Encodes `input` to base64 format.
 * @param input A string.
 * @param opts Encoding options. See {@link Base64Options} and {@link EncodeOptions}.
 * @returns A base64 representation of `input`.
 * @see [Base64Options](Base64Options) for details on properties common to {@link encode} and {@link decode}.
 * @see [EncodeOptions](EncodeOptions) for details on properties exclusive to {@link encode}.
 */
function encode(input, opts) {
    if (opts === void 0) { opts = {}; }
    var Buff = typeof Buffer === 'undefined' ? null : Buffer;
    function isBufferOrString(input) {
        if (typeof input === 'string')
            return true;
        // @ts-expect-error
        if (input instanceof Buff)
            return true;
        return false;
    }
    if (!isBufferOrString(input))
        throw new TypeError("\"input\" must be a string. Got \"".concat(String(input), "\" of type \"").concat(typeof input, "\"."));
    // @ts-expect-error
    var isBuffer = function (input) { return input instanceof Buff; };
    var chars = opts.urlSafe ? URL_SAFE_CHARS : STANDARD_CHARS;
    var output = '';
    var padding = '';
    var inp = isBuffer(input) ? input.toString() : String(input);
    var i = 0;
    while (i < inp.length) {
        var byte1 = inp.charCodeAt(i++) & 0xff;
        var byte2 = inp.charCodeAt(i++) & 0xff;
        var byte3 = inp.charCodeAt(i++) & 0xff;
        var group = (byte1 << 16) | (byte2 << 8) | byte3;
        output +=
            chars[(group >> 18) & 0x3f] +
                chars[(group >> 12) & 0x3f] +
                (byte2 ? chars[(group >> 6) & 0x3f] : '=') +
                (byte3 ? chars[group & 0x3f] : '=');
    }
    var mod = output.length % 4;
    if (mod) {
        padding = '===='.substring(mod);
    }
    return output + padding;
}
exports.encode = encode;
/**
 * Decodes `input` back to its original form.
 * @param input A valid base64 string.
 * @param opts Decoding options. See {@link Base64Options} and {@link DecodeOptions}.
 * @returns The result of decoding `input`.
 * @see [Base64Options](Base64Options) for details on properties common to {@link encode} and {@link decode}.
 * @see [DecodeOptions](DecodeOptions) for details on properties exclusive to {@link decode}.
 */
function decode(input, opts) {
    if (opts === void 0) { opts = {}; }
    // Get the character set to use for decoding
    var chars = opts.urlSafe ? URL_SAFE_CHARS : STANDARD_CHARS;
    // Remove any padding characters from the input string
    var rawInp = (typeof Buffer === 'undefined' ? String(input) : Buffer.from(input).toString());
    var lastEqualIndex = rawInp.lastIndexOf('=');
    var inp = lastEqualIndex !== -1 ? rawInp.substring(0, lastEqualIndex) : rawInp;
    // Split the input string into chunks of 6 bits
    var chunks = map(inp.split(''), function (char) {
        var index = chars.indexOf(char);
        return index >= 0 ? index.toString(2).padStart(6, '0') : '';
    });
    // Combine the chunks into a single binary string
    var binary = chunks.join('');
    // Split the binary string into chunks of 8 bits
    var bytes = binary.match(/.{8}/g) || [];
    // Convert each byte to its corresponding character
    var characters = map(bytes, function (byte) {
        return String.fromCharCode(parseInt(byte, 2));
    });
    // Join the characters to form the final string
    return characters.join('');
}
exports.decode = decode;


cjs/index.d.ts
/// <reference types="node" />
interface Base64Options {
    /**
     * Whether or not to use a URL-safe charset.
     * @since 0.0.3
     */
    urlSafe?: boolean;
}
interface EncodeOptions extends Base64Options {
}
interface DecodeOptions extends Base64Options {
}
/**
 *
 * Encodes `input` to base64 format.
 * @param input A string.
 * @param opts Encoding options. See {@link Base64Options} and {@link EncodeOptions}.
 * @returns A base64 representation of `input`.
 * @see [Base64Options](Base64Options) for details on properties common to {@link encode} and {@link decode}.
 * @see [EncodeOptions](EncodeOptions) for details on properties exclusive to {@link encode}.
 */
declare function encode(input: Buffer | string, opts?: EncodeOptions): string;
/**
 * Decodes `input` back to its original form.
 * @param input A valid base64 string.
 * @param opts Decoding options. See {@link Base64Options} and {@link DecodeOptions}.
 * @returns The result of decoding `input`.
 * @see [Base64Options](Base64Options) for details on properties common to {@link encode} and {@link decode}.
 * @see [DecodeOptions](DecodeOptions) for details on properties exclusive to {@link decode}.
 */
declare function decode(input: string | Buffer, opts?: EncodeOptions): string;
export { encode, decode, Base64Options, DecodeOptions, EncodeOptions };


src/index.ts
interface Base64Options {
	/**
	 * Whether or not to use a URL-safe charset.
	 * @since 0.0.3
	 */
	urlSafe?: boolean;
}
interface EncodeOptions extends Base64Options {
	// Insert any property here
}
interface DecodeOptions extends Base64Options {
	// Insert any property here
}

function createCoder() {
	return new class Coder {
		private readonly opts: Base64Options;
		constructor(opts: Base64Options = {}) {
			this.opts = opts;
		}
		encode(input: string | Buffer) {
			return encode(input, this.opts);
		}
		decode(input: string | Buffer) {
			return decode(input, this.opts);
		}
	};
}

function map<T, R = unknown>(
	array: T[],
	fn: (val: T, index?: number, array?: T[]) => R
) {
	const newValues: R[] = [];

	for (const item of array) {
		newValues.push(fn(item, array.indexOf(item), array));
	}

	return newValues;
}
const URL_SAFE_CHARS =
	'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_';
const STANDARD_CHARS =
	'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
const CHAR_LOOKUP: Record<string, number> = {};
for (let i = 0; i < URL_SAFE_CHARS.length; i++) {
	CHAR_LOOKUP[URL_SAFE_CHARS.charAt(i)] = i;
}
for (let i = 0; i < STANDARD_CHARS.length; i++) {
	CHAR_LOOKUP[STANDARD_CHARS.charAt(i)] = i;
}
/**
 *
 * Encodes `input` to base64 format.
 * @param input A string.
 * @param opts Encoding options. See {@link Base64Options} and {@link EncodeOptions}.
 * @returns A base64 representation of `input`.
 * @see [Base64Options](Base64Options) for details on properties common to {@link encode} and {@link decode}.
 * @see [EncodeOptions](EncodeOptions) for details on properties exclusive to {@link encode}.
 */
function encode(input: Buffer | string, opts: EncodeOptions = {}) {
	const Buff = typeof Buffer === 'undefined' ? null : Buffer;
	function isBufferOrString(input: any) {
		if (typeof input === 'string') return true;
		// @ts-expect-error
		if (input instanceof Buff) return true;

		return false;
	}
	if (!isBufferOrString(input)) 
		throw new TypeError(
			`"input" must be a string. Got "${String(input)}" of type "${typeof input}".`
		);
	// @ts-expect-error
	const isBuffer = (input: any) => input instanceof Buff;
	const chars = opts.urlSafe ? URL_SAFE_CHARS : STANDARD_CHARS;
	let output = '';
	let padding = '';
	const inp = isBuffer(input) ? input.toString() : String(input);
	let i = 0;
	while (i < inp.length) { 
		const byte1 = inp.charCodeAt(i++) & 0xff;
		const byte2 = inp.charCodeAt(i++) & 0xff;
		const byte3 = inp.charCodeAt(i++) & 0xff;
		const group = (byte1 << 16) | (byte2 << 8) | byte3;
		output +=
			chars[(group >> 18) & 0x3f] +
			chars[(group >> 12) & 0x3f] +
			(byte2 ? chars[(group >> 6) & 0x3f] : '=') +
			(byte3 ? chars[group & 0x3f] : '=');
	}
	const mod = output.length % 4;
	if (mod) {
		padding = '===='.substring(mod);
	}
	return output + padding;
}

/**
 * Decodes `input` back to its original form.
 * @param input A valid base64 string.
 * @param opts Decoding options. See {@link Base64Options} and {@link DecodeOptions}.
 * @returns The result of decoding `input`.
 * @see [Base64Options](Base64Options) for details on properties common to {@link encode} and {@link decode}.
 * @see [DecodeOptions](DecodeOptions) for details on properties exclusive to {@link decode}.
 */
function decode(input: string | Buffer, opts: EncodeOptions = {}) {
	// Get the character set to use for decoding
	const chars = opts.urlSafe ? URL_SAFE_CHARS : STANDARD_CHARS;

	// Remove any padding characters from the input string
	
	const rawInp = (typeof Buffer === 'undefined' ? String(input) : Buffer.from(input).toString());
	const lastEqualIndex = rawInp.lastIndexOf('='); 
	const inp = lastEqualIndex !== -1 ? rawInp.substring(0, lastEqualIndex) : rawInp;

	// Split the input string into chunks of 6 bits
	const chunks = map(inp.split(''), (char) => {
		const index = chars.indexOf(char);
		return index >= 0 ? index.toString(2).padStart(6, '0') : '';
	});

	// Combine the chunks into a single binary string
	const binary = chunks.join('');

	// Split the binary string into chunks of 8 bits
	const bytes = binary.match(/.{8}/g) || []; 

	// Convert each byte to its corresponding character
	const characters = map(bytes, (byte) =>
		String.fromCharCode(parseInt(byte, 2))
	);

	// Join the characters to form the final string
	return characters.join('');
}

// The above functions must:
// 1. Be able to handle strings that contain non-printable ASCII (the result of stringifying a
// buffer, for example).
// 2. Be fully ES3 compliant.
// 3. Use padding '=' signs.
// 4. Be fully portable between the browser and Node.js (no reliance in any browser- or Node-exclusive
// API).
// 5. Not modify the argument vector.
// Open an issue or pull request if ANY of the above is not being complied.

export { encode, decode, Base64Options, DecodeOptions, EncodeOptions };
